vita3k/gui/src/controls_dialog.cpp

```diff
@@ -15,45 +15,276 @@
+        host.cfg.keyboard_button_cross = 0;
+    }
+    if (host.cfg.keyboard_button_square < 0 || host.cfg.keyboard_button_square > 231) {
+        host.cfg.keyboard_button_square = 0;
+    }
+    if (host.cfg.keyboard_leftstick_left < 0 || host.cfg.keyboard_leftstick_left > 231) {
+        host.cfg.keyboard_leftstick_left = 0;
+    }
+    if (host.cfg.keyboard_leftstick_right < 0 || host.cfg.keyboard_leftstick_right > 231) {
+        host.cfg.keyboard_leftstick_right = 0;
+    }
+    if (host.cfg.keyboard_leftstick_up < 0 || host.cfg.keyboard_leftstick_up > 231) {
+        host.cfg.keyboard_leftstick_up = 0;
+    }
+    if (host.cfg.keyboard_leftstick_down < 0 || host.cfg.keyboard_leftstick_down > 231) {
+        host.cfg.keyboard_leftstick_down = 0;
+    }
+    if (host.cfg.keyboard_rightstick_left < 0 || host.cfg.keyboard_rightstick_left > 231) {
+        host.cfg.keyboard_rightstick_left = 0;
+    }
+    if (host.cfg.keyboard_rightstick_right < 0 || host.cfg.keyboard_rightstick_right > 231) {
+        host.cfg.keyboard_rightstick_right = 0;
+    }
+    if (host.cfg.keyboard_rightstick_up < 0 || host.cfg.keyboard_rightstick_up > 231) {
+        host.cfg.keyboard_rightstick_up = 0;
+    }
+    if (host.cfg.keyboard_rightstick_down < 0 || host.cfg.keyboard_rightstick_down > 231) {
+        host.cfg.keyboard_rightstick_down = 0;
+    }
+    if (host.cfg.keyboard_button_psbutton < 0 || host.cfg.keyboard_button_psbutton > 231) {
+        host.cfg.keyboard_button_psbutton = 0;
+    }
+}
+
+static void remapper_button(GuiState &gui, HostState &host, int *button) {
+    if (ImGui::Button(SDL_key_to_string[*button])) {
+        gui.oldCapturedKey = *button;
+        gui.isCapturingKeys = true;
+        while (gui.isCapturingKeys == true) {
+            handle_events(host, gui);
+            *button = gui.capturedKey;
+            check_for_sdlrange(gui, host);
+        }
+		config::serialize_config(host.cfg, host.cfg.config_path);
+    }
+}
+
 void draw_controls_dialog(GuiState &gui, HostState &host) {
+    check_for_sdlrange(gui, host);
     float width = ImGui::GetWindowWidth() / 1.35f;
     float height = ImGui::GetWindowHeight() / 1.25f;
     ImGui::SetNextWindowSize(ImVec2(width, height));
     ImGui::SetNextWindowPosCenter();
-    ImGui::Begin("Controls", &gui.help_menu.controls_dialog);
-
-    ImGui::TextColored(GUI_COLOR_TEXT_TITLE, "%-16s    %-16s", "Button", "Mapped button");
-    ImGui::Text("%-16s    %-16s", "Left stick", "WASD");
-    ImGui::Text("%-16s    %-16s", "Right stick", "IJKL");
-    ImGui::Text("%-16s    %-16s", "D-pad", "Arrow keys");
+    ImGui::Begin("Controls", &gui.controls_menu.controls_dialog);
+    ImGui::TextColored(GUI_COLOR_TEXT_MENUBAR_OPTIONS, "%-16s    %-16s", "Button", "Mapped button");
+    ImGui::Text("%-16s", "Left stick up");
+    ImGui::SameLine();
+    ImGui::Dummy(ImVec2(7.0f, 7.0f));
+    ImGui::SameLine();
+    remapper_button(gui, host, &host.cfg.keyboard_leftstick_up);
```

Comment:
Maybe it's worth combining the first few lines into `remapper_button` to reduce repetition? Passing the name (e.x. `"Left stick up"`) as an argument?